File: python/inspect_api/relay.py
from __future__ import annotations

import asyncio
import json
import logging
import threading
import time
from collections import deque
from dataclasses import dataclass
from pathlib import Path
from typing import Any, TextIO

from .constants import RELAY_CAP, REPLAY_CAP, is_error_kind
from .types import Corr, WireEvent

logger = logging.getLogger(__name__)

_MONO_START_NS = time.monotonic_ns()

def _monotonic_offset_ns() -> int:
    return time.monotonic_ns() - _MONO_START_NS

@dataclass
class InspectorSubscription:
    snapshot: list[bytes]
    queue: asyncio.Queue[bytes | None]

class InspectorRelay:
    def __init__(self, session_id: str, session_dir: Path | None) -> None:
        self.session_id = session_id
        self.session_dir = session_dir
        self._seq = 0
        self._lock = threading.Lock()
        self._replay_buffer: deque[bytes] = deque(maxlen=REPLAY_CAP)
        self._ndjson: TextIO | None = None
        self._turn_count: int = 0
        self._last_event_ts: float | None = None
        self._dropped_count: int = 0
        self._turn_id: str | None = None
        self._item_id: str | None = None
        self._response_id: str | None = None
        self._phrase_id: str | None = None
        self._subscribers: list[asyncio.Queue[bytes | None]] = []
        self._sub_loops: list[asyncio.AbstractEventLoop] = []
        self._closed = False

        if session_dir is not None:
            try:
                session_dir.mkdir(parents=True, exist_ok=True)
                ndjson_path = session_dir / f"{session_id}.ndjson"
                self._ndjson = open(ndjson_path, "ab")
            except OSError as err:
                logger.warning("open inspector ndjson failed: %s", err)
                self._ndjson = None

    def next_seq(self) -> int:
        with self._lock:
            s = self._seq
            self._seq += 1
            return s

    def turn_count(self) -> int:
        with self._lock:
            return self._turn_count

    def last_event_ts(self) -> float | None:
        with self._lock:
            return self._last_event_ts

    def dropped_count(self) -> int:
        with self._lock:
            return self._dropped_count

    def corr(self) -> Corr:
        with self._lock:
            return Corr(
                turn_id=self._turn_id,
                item_id=self._item_id,
                response_id=self._response_id,
                phrase_id=self._phrase_id,
            )

    def set_turn_id(self, v: str | None) -> None:
        with self._lock:
            self._turn_id = v

    def set_item_id(self, v: str | None) -> None:
        with self._lock:
            self._item_id = v

    def set_response_id(self, v: str | None) -> None:
        with self._lock:
            self._response_id = v

    def set_phrase_id(self, v: str | None) -> None:
        with self._lock:
            self._phrase_id = v

    def subscriber_count(self) -> int:
        with self._lock:
            return len(self._subscribers)

    def subscribe(self, loop: asyncio.AbstractEventLoop | None = None) -> InspectorSubscription:
        if loop is None:
            try:
                loop = asyncio.get_running_loop()
            except RuntimeError as err:
                raise RuntimeError(
                    "InspectorRelay.subscribe() requires a running asyncio event loop; "
                    "call from an async context or pass loop= explicitly"
                ) from err
        queue: asyncio.Queue[bytes | None] = asyncio.Queue(maxsize=RELAY_CAP)
        with self._lock:
            snapshot = list(self._replay_buffer)
            self._subscribers.append(queue)
            self._sub_loops.append(loop)
        return InspectorSubscription(snapshot=snapshot, queue=queue)

    def unsubscribe(self, queue: asyncio.Queue[bytes | None]) -> None:
        with self._lock:
            for i, q in enumerate(self._subscribers):
                if q is queue:
                    del self._subscribers[i]
                    del self._sub_loops[i]
                    break

    def publish(
        self,
        lane: str,
        kind: str,
        corr_override: Corr | None,
        payload: dict[str, Any],
    ) -> None:
        ts_wall = time.time()
        ts_mono_ns = _monotonic_offset_ns()
        seq = self.next_seq()
        merged = self._merge_corr(corr_override)

        try:
            from otel import current_span_id_hex  # type: ignore[import-not-found]
            span_id = current_span_id_hex()
        except Exception:
            span_id = None

        event = WireEvent(
            session_id=self.session_id,
            seq=seq,
            ts_mono_ns=ts_mono_ns,
            ts_wall=ts_wall,
            lane=lane,
            kind=kind,
            corr=merged,
            span_id=span_id,
            payload=dict(payload),
        )

        try:
            line = json.dumps(event.to_dict(), separators=(",", ":")).encode("utf-8") + b"\n"
        except (TypeError, ValueError) as err:
            logger.warning("serialize inspector event: %s", err)
            return

        emit_mirror = lane != "error" and is_error_kind(kind)

        with self._lock:
            if lane == "turn" and kind == "turn_end":
                self._turn_count += 1
            self._last_event_ts = ts_wall
            self._replay_buffer.append(line)
            if self._ndjson is not None:
                try:
                    self._ndjson.write(line)
                    self._ndjson.flush()
                except OSError as err:
                    logger.warning("write inspector ndjson: %s", err)
            subs = list(zip(self._subscribers, self._sub_loops, strict=True))

        self._fanout(line, subs)

        if emit_mirror:
            mirror_event = self._build_error_mirror(event)
            mirror_event.seq = self.next_seq()
            try:
                mline = (
                    json.dumps(mirror_event.to_dict(), separators=(",", ":")).encode("utf-8")
                    + b"\n"
                )
            except (TypeError, ValueError):
                return
            with self._lock:
                self._replay_buffer.append(mline)
                if self._ndjson is not None:
                    try:
                        self._ndjson.write(mline)
                        self._ndjson.flush()
                    except OSError:
                        pass
                subs = list(zip(self._subscribers, self._sub_loops, strict=True))
            self._fanout(mline, subs)

    def _fanout(
        self,
        line: bytes,
        subs: list[tuple[asyncio.Queue[bytes | None], asyncio.AbstractEventLoop]],
    ) -> None:
        for q, loop in subs:
            try:
                if loop.is_running():
                    loop.call_soon_threadsafe(self._queue_put_nowait, q, line)
                else:
                    try:
                        q.put_nowait(line)
                    except asyncio.QueueFull:
                        with self._lock:
                            self._dropped_count += 1
            except Exception:
                with self._lock:
                    self._dropped_count += 1

    @staticmethod
    def _queue_put_nowait(q: asyncio.Queue[bytes | None], line: bytes) -> None:
        try:
            q.put_nowait(line)
        except asyncio.QueueFull:
            pass

    def _merge_corr(self, override: Corr | None) -> Corr:
        base = self.corr()
        if override is None:
            return base
        return Corr(
            turn_id=override.turn_id if override.turn_id is not None else base.turn_id,
            item_id=override.item_id if override.item_id is not None else base.item_id,
            response_id=override.response_id if override.response_id is not None else base.response_id,
            phrase_id=override.phrase_id if override.phrase_id is not None else base.phrase_id,
        )

    def _build_error_mirror(self, origin: WireEvent) -> WireEvent:
        error_text_raw = origin.payload.get("error") or origin.payload.get("reason")
        if isinstance(error_text_raw, str):
            error_text = error_text_raw
        elif error_text_raw is None:
            error_text = origin.kind
        else:
            error_text = json.dumps(error_text_raw)

        payload: dict[str, Any] = {
            "lane": origin.lane,
            "origin_seq": origin.seq,
            "origin_kind": origin.kind,
            "error": error_text,
            "severity": "error",
        }
        return WireEvent(
            session_id=origin.session_id,
            seq=0,
            ts_mono_ns=origin.ts_mono_ns,
            ts_wall=origin.ts_wall,
            lane="error",
            kind="raised",
            corr=Corr(
                turn_id=origin.corr.turn_id,
                item_id=origin.corr.item_id,
                response_id=origin.corr.response_id,
                phrase_id=origin.corr.phrase_id,
            ),
            span_id=origin.span_id,
            payload=payload,
        )

    def close(self) -> None:
        with self._lock:
            if self._closed:
                return
            self._closed = True
            if self._ndjson is not None:
                try:
                    self._ndjson.flush()
                    self._ndjson.close()
                except OSError:
                    pass
                self._ndjson = None
            subs = list(zip(self._subscribers, self._sub_loops, strict=True))
            self._subscribers.clear()
            self._sub_loops.clear()
        for q, loop in subs:
            try:
                if loop.is_running():
                    loop.call_soon_threadsafe(self._queue_put_nowait, q, None)
                else:
                    try:
                        q.put_nowait(None)
                    except asyncio.QueueFull:
                        pass
            except Exception:
                pass

File: rust/src/inspect/relay.rs
#![allow(dead_code)]

use std::collections::BTreeMap;
use std::fs::{File, OpenOptions};
use std::io::Write;
use std::path::PathBuf;
use std::sync::atomic::{AtomicU64, Ordering};
use std::sync::Mutex;
use std::time::SystemTime;

use serde_json::Value;
use tokio::sync::broadcast;
use tracing::warn;

use super::constants::is_error_kind;
use super::types::{Corr, WireEvent};
use crate::defaults;
use crate::otel;

const FANOUT_CAP: usize = defaults::inspector::RELAY_CAP as usize;
const REPLAY_CAP: usize = (defaults::inspector::RELAY_CAP * 4) as usize;

pub struct InspectorRelay {
    pub session_id: String,
    pub session_dir: Option<PathBuf>,
    seq: AtomicU64,
    inner: Mutex<RelayInner>,
    tx: broadcast::Sender<Vec<u8>>,
}

struct RelayInner {
    replay_buffer: std::collections::VecDeque<Vec<u8>>,
    ndjson: Option<File>,
    turn_count: u64,
    last_event_ts: Option<f64>,
    dropped_count: u64,
    turn_id: Option<String>,
    item_id: Option<String>,
    response_id: Option<String>,
    phrase_id: Option<String>,
}

impl InspectorRelay {
    pub fn new(session_id: String, session_dir: Option<PathBuf>) -> Self {
        let ndjson = session_dir.as_ref().and_then(|dir| {
            let _ = std::fs::create_dir_all(dir);
            let ndjson_path = dir.join(format!("{}.ndjson", session_id));
            OpenOptions::new()
                .create(true)
                .append(true)
                .open(&ndjson_path)
                .map_err(|err| {
                    warn!(error = %err, path = %ndjson_path.display(), "open inspector ndjson failed");
                    err
                })
                .ok()
        });
        let (tx, _rx) = broadcast::channel::<Vec<u8>>(FANOUT_CAP);
        Self {
            session_id,
            session_dir,
            seq: AtomicU64::new(0),
            inner: Mutex::new(RelayInner {
                replay_buffer: std::collections::VecDeque::with_capacity(REPLAY_CAP),
                ndjson,
                turn_count: 0,
                last_event_ts: None,
                dropped_count: 0,
                turn_id: None,
                item_id: None,
                response_id: None,
                phrase_id: None,
            }),
            tx,
        }
    }

    pub fn next_seq(&self) -> u64 {
        self.seq.fetch_add(1, Ordering::Relaxed)
    }

    pub fn turn_count(&self) -> u64 {
        self.inner.lock().expect("relay poisoned").turn_count
    }

    pub fn last_event_ts(&self) -> Option<f64> {
        self.inner.lock().expect("relay poisoned").last_event_ts
    }

    pub fn dropped_count(&self) -> u64 {
        self.inner.lock().expect("relay poisoned").dropped_count
    }

    pub fn corr(&self) -> Corr {
        let g = self.inner.lock().expect("relay poisoned");
        Corr {
            turn_id: g.turn_id.clone(),
            item_id: g.item_id.clone(),
            response_id: g.response_id.clone(),
            phrase_id: g.phrase_id.clone(),
        }
    }

    pub fn set_turn_id(&self, v: Option<String>) {
        self.inner.lock().expect("relay poisoned").turn_id = v;
    }
    pub fn set_item_id(&self, v: Option<String>) {
        self.inner.lock().expect("relay poisoned").item_id = v;
    }
    pub fn set_response_id(&self, v: Option<String>) {
        self.inner.lock().expect("relay poisoned").response_id = v;
    }
    pub fn set_phrase_id(&self, v: Option<String>) {
        self.inner.lock().expect("relay poisoned").phrase_id = v;
    }

    pub fn subscriber_count(&self) -> usize {
        self.tx.receiver_count()
    }

    pub fn subscribe(&self) -> InspectorSubscription {
        let snapshot = {
            let g = self.inner.lock().expect("relay poisoned");
            g.replay_buffer.iter().cloned().collect::<Vec<_>>()
        };
        InspectorSubscription {
            snapshot,
            rx: self.tx.subscribe(),
        }
    }

    pub fn publish(
        &self,
        lane: &str,
        kind: &str,
        corr_override: Option<Corr>,
        payload: BTreeMap<String, Value>,
    ) {
        let now = SystemTime::now();
        let ts_wall = now
            .duration_since(SystemTime::UNIX_EPOCH)
            .map(|d| d.as_secs_f64())
            .unwrap_or(0.0);
        let ts_mono_ns = ts_wall_to_monotonic_ns();
        let seq = self.next_seq();
        let merged_corr = self.merge_corr(corr_override);

        let event = WireEvent {
            session_id: self.session_id.clone(),
            seq,
            ts_mono_ns,
            ts_wall,
            lane: lane.to_string(),
            kind: kind.to_string(),
            corr: merged_corr,
            span_id: otel::current_span_id_hex(),
            payload,
        };

        let mut line = match serde_json::to_vec(&event) {
            Ok(b) => b,
            Err(err) => {
                warn!(error = %err, "serialize inspector event");
                return;
            }
        };
        line.push(b'\n');

        let mirror = if lane != "error" && is_error_kind(kind) {
            Some(self.build_error_mirror(&event))
        } else {
            None
        };

        {
            let mut g = self.inner.lock().expect("relay poisoned");
            if lane == "turn" && kind == "turn_end" {
                g.turn_count += 1;
            }
            g.last_event_ts = Some(ts_wall);
            push_replay(&mut g.replay_buffer, line.clone());
            if let Some(fh) = g.ndjson.as_mut() {
                if let Err(err) = fh.write_all(&line) {
                    warn!(error = %err, "write inspector ndjson");
                }
            }
        }
        let _ = self.tx.send(line);

        if let Some(mirror_event) = mirror {
            let mirror_seq = self.next_seq();
            let mirror_event = WireEvent {
                seq: mirror_seq,
                ..mirror_event
            };
            if let Ok(mut mline) = serde_json::to_vec(&mirror_event) {
                mline.push(b'\n');
                let mut g = self.inner.lock().expect("relay poisoned");
                push_replay(&mut g.replay_buffer, mline.clone());
                if let Some(fh) = g.ndjson.as_mut() {
                    let _ = fh.write_all(&mline);
                }
                drop(g);
                let _ = self.tx.send(mline);
            }
        }
    }

    fn merge_corr(&self, override_corr: Option<Corr>) -> Corr {
        let base = self.corr();
        match override_corr {
            None => base,
            Some(o) => Corr {
                turn_id: o.turn_id.or(base.turn_id),
                item_id: o.item_id.or(base.item_id),
                response_id: o.response_id.or(base.response_id),
                phrase_id: o.phrase_id.or(base.phrase_id),
            },
        }
    }

    fn build_error_mirror(&self, origin: &WireEvent) -> WireEvent {
        let mut payload: BTreeMap<String, Value> = BTreeMap::new();
        payload.insert("lane".into(), Value::String(origin.lane.clone()));
        payload.insert("origin_seq".into(), Value::from(origin.seq));
        payload.insert("origin_kind".into(), Value::String(origin.kind.clone()));
        let error_text = origin
            .payload
            .get("error")
            .or_else(|| origin.payload.get("reason"))
            .map(|v| match v {
                Value::String(s) => s.clone(),
                other => other.to_string(),
            })
            .unwrap_or_else(|| origin.kind.clone());
        payload.insert("error".into(), Value::String(error_text));
        payload.insert("severity".into(), Value::String("error".into()));
        WireEvent {
            session_id: origin.session_id.clone(),
            seq: 0,
            ts_mono_ns: origin.ts_mono_ns,
            ts_wall: origin.ts_wall,
            lane: "error".into(),
            kind: "raised".into(),
            corr: origin.corr.clone(),
            span_id: origin.span_id.clone(),
            payload,
        }
    }

    pub fn close(&self) {
        let mut g = self.inner.lock().expect("relay poisoned");
        if let Some(fh) = g.ndjson.as_mut() {
            let _ = fh.flush();
        }
        g.ndjson = None;
    }
}

pub struct InspectorSubscription {
    pub snapshot: Vec<Vec<u8>>,
    pub rx: broadcast::Receiver<Vec<u8>>,
}

fn push_replay(buf: &mut std::collections::VecDeque<Vec<u8>>, line: Vec<u8>) {
    if buf.len() >= REPLAY_CAP {
        buf.pop_front();
    }
    buf.push_back(line);
}

fn ts_wall_to_monotonic_ns() -> u128 {
    use std::time::Instant;
    static START: std::sync::OnceLock<Instant> = std::sync::OnceLock::new();
    let start = *START.get_or_init(Instant::now);
    start.elapsed().as_nanos()
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    fn temp_dir() -> PathBuf {
        let p = std::env::temp_dir().join(format!(
            "speaches-plus-relay-{}-{}",
            std::process::id(),
            uuid::Uuid::new_v4().simple()
        ));
        std::fs::create_dir_all(&p).unwrap();
        p
    }

    #[test]
    fn publish_writes_ndjson_and_replays_to_subscriber() {
        let dir = temp_dir();
        let relay = InspectorRelay::new("sess_test".into(), Some(dir.clone()));
        let mut payload = BTreeMap::new();
        payload.insert("audio_start_ms".into(), json!(100));
        relay.publish("vad", "confirmed_start", None, payload);

        let path = dir.join("sess_test.ndjson");
        let body = std::fs::read_to_string(&path).unwrap();
        assert!(body.contains("\"lane\":\"vad\""));
        assert!(body.contains("\"kind\":\"confirmed_start\""));
        assert!(body.contains("\"audio_start_ms\":100"));

        let sub = relay.subscribe();
        assert_eq!(sub.snapshot.len(), 1);
        let _ = std::fs::remove_dir_all(&dir);
    }

    #[test]
    fn error_mirror_emits_when_kind_is_in_err_set() {
        let dir = temp_dir();
        let relay = InspectorRelay::new("sess_err".into(), Some(dir.clone()));
        let mut payload = BTreeMap::new();
        payload.insert("error".into(), json!("kaboom"));
        relay.publish("llm", "failed", None, payload);

        let path = dir.join("sess_err.ndjson");
        let body = std::fs::read_to_string(&path).unwrap();
        let lines: Vec<&str> = body.lines().collect();
        assert_eq!(lines.len(), 2, "expected origin + mirror");
        assert!(lines[0].contains("\"lane\":\"llm\""));
        assert!(lines[1].contains("\"lane\":\"error\""));
        assert!(lines[1].contains("\"origin_kind\":\"failed\""));
        assert!(lines[1].contains("\"kaboom\""));
        let _ = std::fs::remove_dir_all(&dir);
    }

    #[test]
    fn corr_merges_with_override() {
        let dir = temp_dir();
        let relay = InspectorRelay::new("sess_corr".into(), Some(dir.clone()));
        relay.set_turn_id(Some("t1".into()));
        relay.set_item_id(Some("i1".into()));
        let corr_override = Corr {
            item_id: Some("i_override".into()),
            ..Default::default()
        };
        relay.publish(
            "turn",
            "user_committed",
            Some(corr_override),
            BTreeMap::new(),
        );

        let path = dir.join("sess_corr.ndjson");
        let body = std::fs::read_to_string(&path).unwrap();
        assert!(body.contains("\"turn_id\":\"t1\""));
        assert!(body.contains("\"item_id\":\"i_override\""));
        let _ = std::fs::remove_dir_all(&dir);
    }

    #[test]
    fn turn_count_tracks_turn_end_kind() {
        let dir = temp_dir();
        let relay = InspectorRelay::new("sess_turn".into(), Some(dir.clone()));
        assert_eq!(relay.turn_count(), 0);
        relay.publish("turn", "turn_start", None, BTreeMap::new());
        assert_eq!(relay.turn_count(), 0);
        relay.publish("turn", "turn_end", None, BTreeMap::new());
        assert_eq!(relay.turn_count(), 1);
        relay.publish("turn", "turn_end", None, BTreeMap::new());
        assert_eq!(relay.turn_count(), 2);
        let _ = std::fs::remove_dir_all(&dir);
    }

    #[test]
    fn subscriber_gets_replay_snapshot_first() {
        let dir = temp_dir();
        let relay = InspectorRelay::new("sess_replay".into(), Some(dir.clone()));
        for i in 0..3 {
            let mut p = BTreeMap::new();
            p.insert("i".into(), json!(i));
            relay.publish("turn", "turn_start", None, p);
        }
        let sub = relay.subscribe();
        assert_eq!(sub.snapshot.len(), 3);
        let _ = std::fs::remove_dir_all(&dir);
    }

    #[test]
    fn replay_buffer_caps_at_replay_cap() {
        let dir = temp_dir();
        let relay = InspectorRelay::new("sess_cap".into(), Some(dir.clone()));
        for _ in 0..(REPLAY_CAP + 100) {
            relay.publish("turn", "turn_start", None, BTreeMap::new());
        }
        let sub = relay.subscribe();
        assert_eq!(sub.snapshot.len(), REPLAY_CAP);
        let _ = std::fs::remove_dir_all(&dir);
    }
}
